### DISCUSSION TITLE
What is more performant, a single Deck instance or multiple per map?

### DISCUSSION THREAD

**Discussion by tomeronen** [visgl/deck.gl#7412] [Q&A] [answered]
I am using DeckGL with react, and i had a question about best practices considering the amount of Deck instances. One option is to add all layers in a single Deck instance, for example:

<DeckGL layers={[layerA, layerB, layerC, layerD]} />
Or to split them between diffrent instances, for example:

<DeckGL layers={[layerA, layerB]} />
<DeckGL layers={[layerC, layerD]} />

The second approach lets you incapsulate diffrent deck instances:

const RegionsLayer = () => <DeckGL layers={[layerA, layerB]} />;
const WeatherLayer = () => <DeckGL layers={[layerC, layerD]} />;
does one approach have a performance advantage compared to the other?

I did both approaches, in basic usage both work well.

**Comment by ibgreen** [comment, marked as answer]
There are no practical advantages to having two DeckGL instances other than if you absolutely need to draw into two separate canvases. Not that deck.gl has a view system to let you draw multiple views into a single canvas.

You'll avoid duplicated overhead with a single deck.gl instance (= single WebGL context), both in terms of GPU and CPU memory as well as browser resources. Things like offscreen GPU buffers for picking etc add to that overhead.

The performance and memory savings of using two views in a single deck instance become more significant if you share layers between the two instances. Less pronounced if each instance shows independent data and layers..

Also be aware that browsers have a limited number of WebGL contexts, so there is a fairly small limit to how many Deck instances you can have.

**Reply by tomeronen** [reply]
thank you so much for the response! if the layers are completely separate in data. Is there a chance that having multiple WebGL contexts will be an advantage? for example that getting more researches from the browser will help with performance?

**Reply by ibgreen** [reply]
WebGL is complex enough that I would never swear that one solution is more performant than another without actually measuring first, but having written a good chunk of the WebGL code in deck.gl, I am not aware of any advantages and can only see disadvantages, and I would try hard to avoid multiple instances in my own apps.

**Reply by tomeronen** [reply]
got you :) again thank you so much for the help
